Enabling worker indexing for Logic Apps app kind behind an enviornment setting

apranaseth opened the pull request:
This is to enable worker indexing for Logic Apps workflow app kind behind an environment variable. The settings used by Logic Apps are "FUNCTIONS_INPROC_NET8_ENABLED": "1", "FUNCTIONS_WORKER_RUNTIME": "dotnet" and we are bringing up a "dotnet" worker.


resolves #11386

### Pull request checklist

**IMPORTANT**: Currently, changes must be backported to the `in-proc` branch to be included in Core Tools and non-Flex deployments.

* [ ] Backporting to the `in-proc` branch is not required
    * Otherwise: https://github.com/Azure/azure-functions-host/pull/11377
* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [x] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [x] I have added all required tests (Unit tests, E2E tests)

<!-- Optional: delete if not applicable  -->
### Additional information

Additional PR information

apranaseth committed:
Enabling worker indexing for Logic Apps app kind behind an enviornment setting.

apranaseth committed:
Enabling worker indexing for Logic Apps app kind behind an enviornment setting.

surgupta-msft commented:
I have a few questions/comments related to this PR -

- Do we need these changes in `dev` branch? My understanding is that Logic apps uses `in-proc` branch only. We may need for consistency purpose between `dev` and `in-proc` - but double checking this.
- Could you please complete the PR checklist in the description and add to the release notes?
- It would be helpful if you can create a github issue and provide context about the changes. I am not very clear about the ask and changes here. Please link github issue in the PR description.
- Please add some tests too.

Let me know if you need help with any of these items, happy to assist.

surgupta-msft commented on src/WebJobs.Script/Utility.cs:
```diff
@@ -1034,7 +1034,8 @@ public static void ValidateRetryOptions(RetryOptions
         public static bool CanWorkerIndex(IEnumerable<RpcWorkerConfig> workerConfigs, IEnvironment environment, FunctionsHostingConfigOptions functionsHostingConfigOptions)
         {
-            if (environment.IsLogicApp())
+            // NOTE(apseth): Enabling the worker indexing for Logic Apps with codeful mode enabled.
```
I think we don't need to add the alias here. We should be able to use the git history to get the info.

RohitRanjanMS replied:
Hi @apranaseth , can you please remove your alias as suggested by @surgupta-msft ?

apranaseth commented:
> I have a few questions/comments related to this PR -
> 
> * Do we need these changes in `dev` branch? My understanding is that Logic apps uses `in-proc` branch only. We may need for consistency purpose between `dev` and `in-proc` - but double checking this.
> * Could you please complete the PR checklist in the description and add to the release notes?
> * It would be helpful if you can create a github issue and provide context about the changes. I am not very clear about the ask and changes here. Please link github issue in the PR description.
> * Please add some tests too.
> 
> Let me know if you need help with any of these items, happy to assist.

Yeah we plan to use the out of proc model soon so wanted to keep these changes in sync. I have updated the checklist and added github issue and test.

surgupta-msft commented on src/WebJobs.Script/Environment/EnvironmentExtensions.cs:
```diff
@@ -462,6 +462,15 @@ public static bool IsLogicApp(this IEnvironment environment)
+        /// </summary>
+        public static bool IsLogicAppCodefulModeEnabled(this IEnvironment environment)
+        {
+            bool.TryParse(environment.GetEnvironmentVariable(EnvironmentSettingNames.LogicAppCodefulModeEnabled), out bool logicAppCodefulModeEnabled);
```
Which component will be setting this Environment variable?

apranaseth replied:
It will be set by the Logic Apps customers in the app settings manually for now. But later we will make it first class.

surgupta-msft commented on src/WebJobs.Script/Environment/EnvironmentSettingNames.cs:
```diff
@@ -155,6 +155,8 @@ public static class EnvironmentSettingNames
 
         public const string AppKind = "APP_KIND";
```
Remove empty line 157 and add a comment for LA settings -

```csharp
// Logic apps settings
public const string AppKind = "APP_KIND";
public const string LogicAppCodefulModeEnabled = "WORKFLOW_CODEFUL_ENABLED";
```

surgupta-msft commented on src/WebJobs.Script/Host/ScriptHost.cs:
```diff
@@ -803,7 +803,7 @@ internal async Task<Collection<FunctionDescriptor>> GetFunctionDescriptorsAsync(
             if (!cancellationToken.IsCancellationRequested)
             {
-                bool throwOnWorkerRuntimeAndPayloadMetadataMismatch = true;
+                bool throwOnWorkerRuntimeAndPayloadMetadataMismatch = !(_environment.IsLogicApp() && _environment.IsLogicAppCodefulModeEnabled());
```
Not clear on why do we need to make changes here?

apranaseth replied:
Since Logic Apps is using dotnet as the runtime, if you look below this is set to true by default and we would throw an exception here which causes the host startup to fail.

surgupta-msft commented on src/WebJobs.Script/Utility.cs:
For scenario, `if (environment.IsLogicApp() && environment.IsLogicAppCodefulModeEnabled())`, below line will be executed -

```csharp
bool workerIndexingDisabledViaHostingConfig = functionsHostingConfigOptions.WorkerIndexingDisabledApps.ToLowerInvariant().Split("|").Contains(appName);
```

This line is for Functions apps for which indexing is disabled via hosting config. Let's say indexing is disabled for function app "app1" and there is a logic app with the same name, then indexing will be disabled for that LA too. But our requirement is to enable indexing for above scenario.

Fix -

```csharp
bool workerIndexingDisabledViaHostingConfig = !environment.IsLogicApp() && functionsHostingConfigOptions.WorkerIndexingDisabledApps.ToLowerInvariant().Split("|").Contains(appName);
```

Add/update existing tests to cover this scenario.

apranaseth replied:
A Logic App and a function app can't have same names. As end of the day its a web app and web apps names should be unique.

apranaseth replied:
Since app names are unique, we dont need to handle this separately for Logic Apps case.

apranaseth committed:
Disabled multi language workers for codeful mode.

soninaren reviewed (changes requested):
Would be good to add log statements for better tracking. as mentioned in the in-proc PR.

RohitRanjanMS commented on test/WebJobs.Script.Tests/UtilityTests.cs:
```diff
@@ -960,6 +960,31 @@ private void VerifyCanWorkerIndexUtility(bool workerIndexingFeatureFlag, bool wo
+        [InlineData(true, false, false)]
+        public void WorkerIndexingDecisionLogic_LogicApps(bool workerIndexingFeatureFlag, bool enableIndexingForCodeful, bool expected)
+        {
+            var testEnv = new TestEnvironment();
```
See if you can use TestScopedEnvironmentVariable, this will clean-up the variable. You will see example of this in other tests.

apranaseth replied:
I have cleaned up the testenv to clean the set variables.

RohitRanjanMS commented on test/WebJobs.Script.Tests/UtilityTests.cs:
```diff
@@ -960,6 +960,31 @@ private void VerifyCanWorkerIndexUtility(bool workerIndexingFeatureFlag, bool wo
+
+            if (enableIndexingForCodeful)
+            {
+                testEnv.SetEnvironmentVariable(EnvironmentSettingNames.LogicAppCodefulModeEnabled, "true");
```
We should cleanup the env variables.

apranaseth replied:
I have cleaned up the testenv to clean the set variables.

apranaseth committed:
Cleaned up test and added release notes.

apranaseth commented:
> Would be good to add log statements for better tracking. as mentioned in the in-proc PR.

As we discussed we will log this setting in the Logic Apps extension
